### DISCUSSION TITLE
docs/backend/Java/reactive-programming/yi-wen-che-di-gao-dong-webflux-zhong-de-publishonhe-subscribeon-zen-me-shi-yong

### DISCUSSION THREAD

**Discussion by giscus** [MingGH/runnable-run-giscus#2] [General]
# docs/backend/Java/reactive-programming/yi-wen-che-di-gao-dong-webflux-zhong-de-publishonhe-subscribeon-zen-me-shi-yong

1. 前言

http://localhost:8081/docs/backend/Java/reactive-programming/yi-wen-che-di-gao-dong-webflux-zhong-de-publishonhe-subscribeon-zen-me-shi-yong

**Comment by fattyca1** [comment]
订阅内部流的执行在其他线程, 且设置flatMap的并行度为2，但是整个流还是同步执行，因为
Mono.just 为意味着热数据，每次会同步的执行去获取，因此执行的时候仍然为同步。  

这句话要怎么理解呢？ 没看懂， 如果flatmap中是并行执行的，那为什么不能并行的去执行executeRequest()方法呢？ 
难道是因为subscirbe的时候，只会request 1 ，然后要等到executeRequest方法执行完？  

带佬帮忙解释下， 谢谢

**Reply by MingGH** [reply]
这里有两个点：
- Mono.just 热数据会在组装时就计算好这个值，和是否已经subscribe没关系。看如下代码：
```java
public static void main(String[] args) throws InterruptedException {
    Mono<String> reqMono = Mono.just(executeRequest());

    Thread.sleep(1000);
    System.out.println("等待一秒结束，开始订阅");
    reqMono.doOnNext(it -> {
                System.out.println("得到结果" + it);
            })
            .subscribe();
}


private static String executeRequest(){
    System.out.println("开始请求");
    ...
}

```


在上面的代码中，我在`Mono.just` 中写了个`executeRequest()`，但是等待1秒后才开始subscribe。

执行时会发现`subscribe`前就已经执行`executeRequest()`这个方法了。

- 理解这个之后，回到文章中的那个例子：

```java
.flatMap(it -> Mono.just(executeRequest())
                .subscribeOn(Schedulers.newParallel("REQ-T", 2))
                .log()
, 2)
```

在flatMap执行时，其实就是：flatMap subscribe Mono.just(executeRequest()) 

又因为上面第一点我们说到 `Mono.just(executeRequest())`中对于是否执行executeRequest来说和“**mono有没有已经subscribe**”没关系，而和“**Mono.just(executeRequest())是否已经组装**”有关系，所以即便flatMap，设置了并行度为2，但因为flatMap 的并行性体现是在消费阶段，而`Mono.just(executeRequest())`是一个一个组装的，所以还是串行执行
